### PR TITLE
Proposal: Implement license choices

### PR DESCRIPTION
**Pull request by mnonnenmacher**
A proposal for how license choices could be implemented to collect feedback.

### CODE REVIEW

**Comment by fviernau** [comment]
An slightly different idea for the implementation of license choice came to my head just yesterday.
Let's see an example:
```
license: "Apache-2.0 OR MIT OR GPL-2.0-only"
choice: "Apache.2.0"
```
e.g. the choice is just a string containing and SPDX expression and we can have a function which determines if that string is a valid choice for `license`.

As extra one could think about support narrowing down the choice, e.g. in a global configuration, to limit the choices e.g. inside your org:
```
choice: Apache-2.0 OR MIT
```

**Comment by tsteenbe** [comment]
My thinking has always been to keep things simple and have a `ort-config-path/license-choices.yml` where user could specify "if you find this OR SPDX expression then do this."
Optionally we could also add this to `curations.yml` to allow a global and package-specific license choices.

`licenses-choices.yml` would look like this:

```
- expression: "Apache-2.0 OR MIT"
  choice: "MIT"
  comment "Optional explanation from legal counsel / developer why this choice was made."
- expression: "Apache-2.0 OR GPL-1.0 OR MIT"
  choice: "MIT"
- expression: "MIT"
  choice: "MIT"
- expression: "(Apache-2.0 AND CC0-1.0) OR MIT"
  choice: "CC0-1.0 AND MIT" 
- expression: "Apache-2.0 OR Apache-2.0 WITH LLVM-exception"
  choice: "Apache-2.0" 
```

Pros:
- Simple implementation in ORT - replace every found OR SPDX expression with choice  found in licenses-choices.yml - basic find and replace
Cons:
- User has to specify every possible combination

File: spdx-utils/src/main/kotlin/LicenseChoice.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+package org.ossreviewtoolkit.spdx
+
+/**
+ * This class represents a license choice for [SpdxExpression]s that contain the [OR operator][SpdxOperator.OR].
+ *
+ * The [choice] property contains the chosen license. The [alternatives] list contains a list of alternative licenses
+ * that need to be concatenated with [choice] via [OR][SpdxOperator.OR] for the choice to apply. This is useful to make
+ * sure that a license choice is not automatically applied if the alternative licenses of a package change, or when
+ * making global license choices.
```
Comment (sschuberth) [inline spdx-utils/src/main/kotlin/LicenseChoice.kt:28]: Basically I like the idea of limiting the "scope" of a license choice via "alternatives", but if doing that I'd propose a different name, e.g. "context". The downside of that approach might be that it's a bit technical / harder to grasp than a (potentially bigger) explicit list of mappings.

**Comment by fviernau** [comment]
I recapped my above thoughts and propose that a concluded license consists of a matcher bit, and
of the actual choice, as follows:

1. matcher for package identifier
   a) all packages
   b) packages matching given name type group and optionally given version
2. matcher for license
   a) expression subtree matcher 
   b) whole expression matcher 
3. matcher for license source (optional)
4. The choice, replaces the matched license bit with the actual choice.

**Comment by sschuberth** [comment]
@mnonnenmacher I believe this can be closed now that https://github.com/oss-review-toolkit/ort/pull/3425 is merged, right?

**Comment by mnonnenmacher** [comment]
> @mnonnenmacher I believe this can be closed now that #3425 is merged, right?

Yes.

**Pull request closed without merging**